src/race.rs

```diff
@@ -175,51 +176,275 @@ impl OnceBool {
+
+            let exchange = once_ptr.compare_exchange(
+                ptr::null_mut(),
+                ptr,
+                Ordering::AcqRel,
+                Ordering::Acquire,
+            );
+
+            if let Err(old) = exchange {
+                ptr = old;
+            }
+        }
+
+        Ok(unsafe { NonNull::new_unchecked(ptr) })
+    }
+}
+
+impl<T> OncePointee for [T] {
+    // An uninitialized slice pointer has both a null address and invalid
+    // length. This enables us to initialize the parts in two separate
+    // operations while ensuring that concurrent readers can determine whether
+    // the slice is fully initialized.
+    //
+    // TODO: Use `slice_from_raw_parts_mut` once stable in `const`.
+    const UNINIT: *mut Self = ptr::slice_from_raw_parts(ptr::null::<T>(), usize::MAX) as _;
+
+    fn get(once: &OncePtrInner<Self>) -> *mut Self {
+        let (once_ptr, once_len) = once.atomic_parts();
+
+        let mut ptr = ptr::null_mut();
+
+        let len = once_len.load(Ordering::Acquire);
+        let len_is_init = (len as isize) >= 0;
+
+        // The pointer is not initialized until after the length is.
+        if len_is_init {
+            ptr = once_ptr.load(Ordering::Acquire);
+        }
+
+        ptr::slice_from_raw_parts_mut(ptr, len)
+    }
+
+    fn set(once: &OncePtrInner<Self>, new: NonNull<Self>) -> Result<(), ()> {
+        let new_ptr = new.as_ptr() as *mut T;
+        let new_len = new.len();
+
+        let (once_ptr, once_len) = once.atomic_parts();
+
+        let exchange =
+            once_len.compare_exchange(usize::MAX, new_len, Ordering::AcqRel, Ordering::Acquire);
+
+        match exchange {
+            Ok(_) => {
+                // We initialized our length first, so store our pointer.
+                once_ptr.store(new_ptr, Ordering::Release);
+                Ok(())
+            }
+            Err(_) => Err(()),
+        }
+    }
+
+    fn get_or_try_init<F, E>(once: &OncePtrInner<Self>, f: F) -> Result<NonNull<Self>, E>
+    where
+        F: FnOnce() -> Result<NonNull<Self>, E>,
+    {
+        let (once_ptr, once_len) = once.atomic_parts();
+
+        let mut ptr = once_ptr.load(Ordering::Acquire);
+        let mut len;
+
+        if ptr.is_null() {
+            let slice_ptr = f()?;
+            ptr = slice_ptr.as_ptr() as *mut T;
+            len = slice_ptr.len();
+
+            // Attempt to initialize our length.
+            let exchange =
+                once_len.compare_exchange(usize::MAX, len, Ordering::AcqRel, Ordering::Acquire);
+
+            match exchange {
+                Ok(_) => {
+                    // We initialized our length first, so store our pointer.
+                    once_ptr.store(ptr, Ordering::Release);
+                }
+                Err(real_len) => {
+                    len = real_len;
+
+                    // Spin briefly until the other thread initializes the
+                    // pointer.
+                    ptr = loop {
```

Comment:
It’s a fundamental invariant of OnceCell that it doesn’t do unbounded spinning. If mutex-like blocking behavior is needed, then std dependency is required, because correct blocking fundamentally requires talking to the OS and notifying it which other thread you are blocked on, so that the OS can correctly guard against priority inversion.

Reply:
My understanding is that no good solutions are available here: either a mutex or a cas is required. Though, I think Rust has double-word cas available on at least some architectures:

https://doc.rust-lang.org/stable/core/arch/x86_64/fn.cmpxchg16b.html

So I think the path forward here is to add support for DSTs via double-word-cas, by adding a bunch of cfgs everywhere. Eg, I'd imagine some 32 processors might have AtomicU64? Though, doing that probably requires a research of what's the status quo for wider-than-word atomic operations in today's Rust. That intrinsic exists, so its non-zero, but I don't know best practicies for actually using that.